src/mongocrypt-crypto.c

```diff
@@ -1016,3 +1030,212 @@ _mongocrypt_hmac_sha_256 (_mongocrypt_crypto_t *crypto,
    }
    return _native_crypto_hmac_sha_256 (key, in, out, status);
 }
+
+bool
+_mongocrypt_fle2_do_encryption (_mongocrypt_crypto_t *crypto,
+                                const _mongocrypt_buffer_t *iv,
+                                const _mongocrypt_buffer_t *associated_data,
+                                const _mongocrypt_buffer_t *key,
+                                const _mongocrypt_buffer_t *plaintext,
+                                _mongocrypt_buffer_t *ciphertext,
+                                uint32_t *bytes_written,
+                                mongocrypt_status_t *status)
+{
+   BSON_ASSERT_PARAM (crypto);
+   BSON_ASSERT_PARAM (iv);
+   BSON_ASSERT_PARAM (associated_data);
+   BSON_ASSERT_PARAM (key);
+   BSON_ASSERT_PARAM (plaintext);
+   BSON_ASSERT_PARAM (ciphertext);
+   BSON_ASSERT_PARAM (bytes_written);
+   BSON_ASSERT_PARAM (status);
+
+   if (ciphertext->len !=
+       _mongocrypt_fle2_calculate_ciphertext_len (plaintext->len)) {
+      CLIENT_ERR ("output ciphertext must be allocated with %" PRIu32 " bytes",
+                  _mongocrypt_fle2_calculate_ciphertext_len (plaintext->len));
+      return false;
+   }
+
+   if (plaintext->len <= 0) {
+      CLIENT_ERR ("input plaintext too small. Must be more than zero bytes.");
+      return false;
+   }
+
+   if (MONGOCRYPT_IV_LEN != iv->len) {
+      CLIENT_ERR ("IV must be length %d, but is length %" PRIu32,
+                  MONGOCRYPT_IV_LEN,
+                  iv->len);
+      return false;
+   }
+   if (MONGOCRYPT_KEY_LEN != key->len) {
+      CLIENT_ERR ("key must be length %d, but is length %" PRIu32,
+                  MONGOCRYPT_KEY_LEN,
+                  key->len);
+      return false;
+   }
+
+   memset (ciphertext->data, 0, ciphertext->len);
+   *bytes_written = 0;
+
+   /* Declare variable names matching [AEAD with
+    * CTR](https://docs.google.com/document/d/1eCU7R8Kjr-mdyz6eKvhNIDVmhyYQcAaLtTfHeK7a_vE/).
+    */
+   /* M is the input plaintext. */
+   _mongocrypt_buffer_t M = {.data = plaintext->data, .len = plaintext->len};
```

Comment:
Should probably preserve constness on M/Ke/IV/Km/AD since they come from const sources. Decrypt block as well, though inverted selection, of course.

Reply:
Also, how common is this pattern? Would it be worth making a view API?

```
bool _mongocrypt_buffer_view (_mongocrypt_buffer_t *out, _mongocrypt_buffer_t *in, size_t ofs, size_t len);
bool _mongocrypt_buffer_const_view (const _mongocrypt_buffer_t *out, const _mongocrypt_buffer_t *in, size_t ofs, size_t len);
```

An API would help guard against struct changes and more importantly build bounds checking into the calls to protect against mistakes.  Though I'll be honest, writing that out makes me miss templates and auto. :)

Reply:
Added a `_mongocrypt_buffer_view` to do bounds checking.

I think the constness of the arguments is preserved. `const _mongocrypt_buffer_t` ensures the `data` pointer and `len` are not modified. But it does not prevent modification to the data pointed to by the `data`.

As an aside: I think it would be safer to have the `const _mongocrypt_buffer_t` arguments guarantee that data pointed to by `data` is not modified. One idea is to consider an API similar to the [mstr types](https://github.com/mongodb/libmongocrypt/blob/master/src/mlib/str.h#L14-L86). We could have a separate `_mongocrypt_buffer_view` type containing a `const uint8_t* data` pointer. But I prefer keeping that out of the scope of this PR.

Reply:
Right. Too much C++ in my head.